### PR TITLE
[134] Implement send job offer API @POST

### PR DESCRIPTION
**Pull request by badnameok**
## What did you do

-implement the API which allows casting to send job offer and mark the application as `OFFER_SENT`.

## Demo

- go to swagger
- login as actor and apply for a job
- use lib:studio to see if the application has been created

# Testing success
- login as casting of the respective job and use the new offer API to send offer using the same `actorId` and `jobId`.
- check in prisma studio again if the application status is now `OFFER_SENT`
- try using the api with exact same parameter again, it should become a bad request
# Testing error
- try using arbitrary jobId and see if it's a 404 or forbidden due to not being owner
- try using arbitrary actorId and see if it's a bad request (actor has not applied)

### CODE REVIEW

File: apps/api/src/modules/job/application/offer/offer.module.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import { Module } from '@nestjs/common'
+import { PrismaModule } from 'src/database/prisma.module'
+
+import { ApplicationModule } from '../application.module'
+import { OfferController } from './offer.controller'
+import { OfferRepository } from './offer.repository'
+import { OfferService } from './offer.service'
+
+@Module({
+  imports: [ApplicationModule, PrismaModule],
```
Comment (palmcm) [inline apps/api/src/modules/job/application/offer/offer.module.ts:10]: Why import Application module

File: apps/api/src/modules/job/application/offer/offer.repository.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import { Injectable } from '@nestjs/common'
+import { PrismaService } from 'src/database/prisma.service'
+
+import { ApplicationRepository } from '../application.repository'
+
+@Injectable()
+export class OfferRepository {
+  constructor(
+    private prisma: PrismaService,
+    private appRepo: ApplicationRepository,
```
Comment (palmcm) [inline apps/api/src/modules/job/application/offer/offer.repository.ts:11]: Optional
```suggestion
    private applicationRepository: ApplicationRepository,
```

File: apps/api/src/modules/job/application/offer/offer.controller.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  @Put('/:jobId/actors/:actorId/offer')
+  @UseAuthGuard(UserType.CASTING)
+  @ApiOperation({ summary: 'Sends a job offer to actor' })
+  @ApiOkResponse()
+  @ApiBadRequestResponse({
+    description:
+      'Actor has not applied to this job or an offer has already been sent.',
+  })
+  @ApiUnauthorizedResponse({
+    description: 'User is not logged in or is not a casting',
```
Comment (palmcm) [inline apps/api/src/modules/job/application/offer/offer.controller.ts:31]: not casting should be forbidden

File: apps/api/src/modules/job/application/offer/offer.controller.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  })
+  @ApiForbiddenResponse({ description: 'User is not the owner of this job' })
+  @ApiNotFoundResponse({ description: 'Job not found' })
+  sendJobOffer(
+    @Param('jobId') jobId: number,
+    @Param('actorId') actorId: number,
```
Comment (palmcm) [inline apps/api/src/modules/job/application/offer/offer.controller.ts:37]: usually @Param() will be type string but optional?
```suggestion
    @Param('jobId') jobId: string,
    @Param('actorId') actorId: string,
```

File: apps/api/src/modules/job/application/offer/offer.service.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    if (!job) {
+      throw new NotFoundException('Job not found')
+    }
+    if (job.castingId !== castingId) {
+      throw new ForbiddenException('You are not the owner of this job')
+    }
+    const application = await this.offerRepository.getApplication(
+      actorId,
+      jobId,
+    )
```
Comment (palmcm) [inline apps/api/src/modules/job/application/offer/offer.service.ts:26]: use getApplicationbyActorJob in ApplicationRepository

File: apps/api/src/modules/job/application/offer/offer.repository.ts [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    return await this.prisma.job.findFirst({
+      where: {
+        jobId: jobId,
+      },
+    })
+  }
+
+  async getApplication(actorId: number, jobId: number) {
+    return await this.appRepo.getApplicationbyActorJob(actorId, jobId)
+  }
```
Comment (palmcm) [inline apps/api/src/modules/job/application/offer/offer.repository.ts:39]:
```suggestion
```
just call directly

File: apps/api/src/modules/job/application/offer/offer.repository.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
+import { ApplicationStatus } from '@modela/database'
+import { Injectable } from '@nestjs/common'
+import { PrismaService } from 'src/database/prisma.service'
+
+import { ApplicationRepository } from '../application.repository'
+
+@Injectable()
+export class OfferRepository {
```
Comment (palmcm) [inline apps/api/src/modules/job/application/offer/offer.repository.ts:8]: Writing this after review this file maybe we remove this file

File: apps/api/src/modules/job/application/offer/offer.repository.ts [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    await this.prisma.application.update({
+      where: {
+        applicationId: applicationId,
+      },
+      data: {
+        status: status,
+      },
+    })
+    return
+  }
```
Comment (palmcm) [inline apps/api/src/modules/job/application/offer/offer.repository.ts:27]: I think this should move to applicationRepository

File: apps/api/src/modules/job/application/offer/offer.repository.ts [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+  async getJobById(jobId: number) {
+    return await this.prisma.job.findFirst({
+      where: {
+        jobId: jobId,
+      },
+    })
+  }
```
Comment (palmcm) [inline apps/api/src/modules/job/application/offer/offer.repository.ts:35]: use getBaseJobById in jobRepository
and actually you should use findUnique not findFirst

File: apps/api/src/modules/job/application/offer/offer.service.ts [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+} from '@nestjs/common'
+
+import { OfferRepository } from './offer.repository'
+
+@Injectable()
+export class OfferService {
+  constructor(private readonly offerRepository: OfferRepository) {}
+
+  async sendJobOffer(jobId: number, castingId: number, actorId: number) {
+    const job = await this.offerRepository.getJobById(jobId)
```
Comment (palmcm) [inline apps/api/src/modules/job/application/offer/offer.service.ts:16]: use getBaseJobById in jobRepository

**Review by palmcm — CHANGES_REQUESTED** [review]

File: apps/api/src/modules/job/application/offer/offer.controller.ts [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+
+@ApiTags('offer')
+@Controller('jobs')
+export class OfferController {
+  constructor(private readonly offerService: OfferService) {}
+
+  @Put('/:jobId/actors/:actorId/offer')
+  @UseAuthGuard(UserType.CASTING)
+  @ApiOperation({ summary: 'Sends a job offer to actor' })
+  @ApiOkResponse()
```
Comment (miello) [inline apps/api/src/modules/job/application/offer/offer.controller.ts:25]: Maybe add some description here ?

File: apps/api/src/modules/job/application/offer/offer.controller.ts [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+    description: 'User is not logged in or is not a casting',
+  })
+  @ApiForbiddenResponse({ description: 'User is not the owner of this job' })
+  @ApiNotFoundResponse({ description: 'Job not found' })
+  sendJobOffer(
+    @Param('jobId') jobId: number,
+    @Param('actorId') actorId: number,
+    @User() user: JwtDto,
+  ) {
+    return this.offerService.sendJobOffer(+jobId, user.userId, +actorId)
```
Comment (miello) [inline apps/api/src/modules/job/application/offer/offer.controller.ts:40]:
```suggestion
    return await this.offerService.sendJobOffer(+jobId, user.userId, +actorId)
```

**Review by miello — CHANGES_REQUESTED** [review]

File: apps/api/src/modules/job/application/offer/offer.module.ts (hunk 1 above)
Comment (badnameok) [reply apps/api/src/modules/job/application/offer/offer.module.ts:10]: it is used to get application for the job offer

File: apps/api/src/modules/job/application/offer/offer.repository.ts (hunk 2 above)
Comment (badnameok) [reply apps/api/src/modules/job/application/offer/offer.repository.ts:11]: removed.

File: apps/api/src/modules/job/application/offer/offer.module.ts (hunk 1 above)
Comment (palmcm) [reply apps/api/src/modules/job/application/offer/offer.module.ts:10]: Ah I read it wrong I thought it a AppModule

File: apps/api/src/modules/job/application/offer/offer.repository.ts
Comment (palmcm) [inline apps/api/src/modules/job/application/offer/offer.repository.ts:1]: Delete this file, I think anything in this file will can be put in ApplicationRepository

File: apps/api/src/modules/job/application/offer/offer.module.ts [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+import { JobModule } from '../../job.module'
+import { ApplicationModule } from '../application.module'
+import { OfferController } from './offer.controller'
+import { OfferRepository } from './offer.repository'
+import { OfferService } from './offer.service'
+
+@Module({
+  imports: [ApplicationModule, JobModule, PrismaModule],
+  controllers: [OfferController],
+  providers: [OfferService, OfferRepository],
```
Comment (palmcm) [inline apps/api/src/modules/job/application/offer/offer.module.ts:13]: Don't forget to remove this

**Review by palmcm — APPROVED** [review]

**Review by miello — APPROVED** [review]

**Pull request merged**